src/api/users/user.service.ts

```diff
@@ -0,0 +1,90 @@
+import { HttpException, HttpStatus, Injectable } from "@nestjs/common";
+import { InjectRepository } from "@nestjs/typeorm";
+import { createHash } from "crypto";
+import { sign, decode } from 'jsonwebtoken';
+import { Repository } from "typeorm";
+import { CreateUserDto } from "./dto/createUser.dto";
+import { LoginUserDto } from "./dto/loginUser.dto";
+import { UserResponseInterface } from "./types/userResponse.interface";
+import { User } from "./user.model";
+
+@Injectable()
+export class UserService {
+    constructor(
+        @InjectRepository(User)
+        private readonly userRepository: Repository<User>,
+    ) { }
+
+    //Registration user
+    async createUser(createUserDto: CreateUserDto): Promise<User> {
+        const user = await this.userRepository.findOne({
+            email: createUserDto.email,
+        })
+        if (user) {
+            throw new HttpException('Email are taken', HttpStatus.UNPROCESSABLE_ENTITY);
+        }
+        createUserDto.password = this.hashPassword(createUserDto.password);
+
+        const createUser = new User();
+        Object.assign(createUser, createUserDto);
+
+        return await this.userRepository.save(createUser);
+    }
+
+    async loginUser(loginUserDto: LoginUserDto): Promise<User> {
+        const user = await this.userRepository.findOne(
+            { email: loginUserDto.email },
+            { select: ['id', 'email', 'password', 'firstName', 'lastName', 'clan_id'] },
+        )
+        if (!user) {
+            throw new HttpException('User not found', HttpStatus.NOT_FOUND);
+        }
+
+        loginUserDto.password = this.hashPassword(loginUserDto.password);
+
+        if (loginUserDto.password !== user.password) {
+            throw new HttpException('Invalid password', HttpStatus.BAD_REQUEST);
+        }
+        delete user.password;
+
+        return user;
+    }
+
+
+
+    //Regular functions
+    buildUserResponse(user: User): UserResponseInterface {
+        return {
+            user: {
+                ...user,
+                token: this.generateJwt(user)
+            }
+        }
+    }
+
+    hashPassword(password: string) {
+        return createHash('sha256')
+            .update(`${password}${process.env['PASSWORD_HASH_SALT']}`)
+            .digest()
+            .toString('hex');
+    }
+
+    findById(id: number, options?): Promise<User> {
+        return this.userRepository.findOne(id, options);
+    }
+
+    findByParams(params: object): Promise<User> {
+        return this.userRepository.findOne(params);
+    }
+
+    public generateJwt(user: User): string {
+        return sign({
+            id: user.id,
+            email: user.email,
+        }, process.env['JWT_SECRET'])
```

Comment:
We might need to verify JWT_SECRET and PASSWORD_HASH_SALT on app init level. Otherwise it's possible situation to have inconsitency like: I've tried to create user, but it's failed. I try to register again and get an error with "user with the same email already exists".

Reply:
- TEMPORARILY RESOLVED -